README.md
### kitIdea

kitIdea is an extension for the Customer Relationship Management (CRM) [KeepInTouch] [3] and the Content Management Systems [WebsiteBaker] [1] or [LEPTON CMS] [2]. It enables the flexible creating and sharing of ideas. 

#### Requirements

* minimum PHP 5.2.x
* using [WebsiteBaker] [1] _or_ using [LEPTON CMS] [2]
* [wblib] [10] must be installed
* [LibraryAdmin] [11] must be installed
* [lib_jquery] [12] must be installed
* [dbConnect_LE] [4] must be installed 
* [Dwoo] [5] must be installed
* [DropletsExtension] [6] must be installed
* [KeepInTouch] [3] must be installed
* [kitForm] [13] must be installed
* [kitDirList] [14] must be installed

#### Installation

* download the actual [kitIdea_x.xx.zip] [7] installation archive
* in CMS backend select the file from "Add-ons" -> "Modules" -> "Install module"

#### First Steps

Please visit the [phpManufaktur] [8] to get more informations about **kitIdea** and join the [Addons Support Group] [9].

[1]: http://websitebaker2.org "WebsiteBaker Content Management System"
[2]: http://lepton-cms.org "LEPTON CMS"
[3]: https://addons.phpmanufaktur.de/download.php?file=KeepInTouch
[4]: https://addons.phpmanufaktur.de/download.php?file=dbConnect_LE
[5]: https://addons.phpmanufaktur.de/download.php?file=Dwoo
[6]: https://addons.phpmanufaktur.de/download.php?file=DropletsExtension
[7]: https://addons.phpmanufaktur.de/download.php?file=kitIdea
[8]: http://phpmanufaktur.de/cms/websitebaker/addons/kitidea.php
[9]: https://phpmanufaktur.de/support
[10]: https://github.com/webbird/wblib/downloads
[11]: http://jquery.lepton-cms.org/modules/download_gallery/dlc.php?file=75&id=1318585713
[12]: http://jquery.lepton-cms.org/modules/download_gallery/dlc.php?file=76&id=1320743410
[13]: https://addons.phpmanufaktur.de/download.php?file=kitForm
[14]: https://addons.phpmanufaktur.de/download.php?file=kitDirList


CHANGELOG
### kitIdea
(c) 2011 - 2012 by phpManufaktur
Ralf Hertsch, Stockholmer Str. 31, 13359 Berlin (Germany)
http://phpManufaktur.de - ralf.hertsch@phpManufaktur.de

Release 0.28 - 2012-10-05

* changed implementation of the Dwoo template engine
* separated cronjob from KIT and add a own cronjob.php

Release 0.27 - 2012-06-19

* added CHANGELOG
* removed info.txt 
* changed WebsiteBaker version check
* updated header informations
* changed license from GNU GPL to MIT
* added LICENSE
* added README.md

Release 0.26 - 2012-02-24

* added: Dwoo plugin translate() for direct I18n translations within the templates
* added: option to switch of the WYSIWYG editor if not really needed
* changed template: please read /modules/kit_idea/templates/changes.txt for 
  more informations! 
* changed: install.php, upgrade.php - removed switching to about dialog (caused 
  problems at some installations)

Release 0.25 - 2012-02-14

* changed some small typos.
* fixed: standard section name can contain leading space, which causes problems
* changed: introduce $_SESSION['KIT_HTML_REQUEST'] to mark fields which should 
  not processed by xssPrevent()
* added: check the language files for correct UTF-8 encoding
* changed template: please read /modules/kit_idea/templates/changes.txt for
  more informations! 

Release 0.24 - 2012-02-01

* fixed: the cronjob does not check if the mailing function is switched off
* added: 'user_status' enable the user to see that they are logged in and to see
  and set the email information mode.
* added: parameter 'user_status' = true|false in droplet kit_idea, default=false
* added template: please read /modules/kit_idea/templates/changes.txt for
  more informations!

Release 0.23 - 2012-01-31

* added: projects can now moved between the different project groups
* changed templates: please read /modules/kit_idea/templates/changes.txt for
  more informations!

Release 0.22 - 2011-11-27

* added: the used toolbar of the WYSIWYG editor now depends on the permission
  article_edit_html, so admins may get more rights and abilities
* changed: it is now possible to change "project" and "projects" to other terms
* fixed: XHTML error in in /modules/kit_idea/templates/1/DE/project.overview.lte
* added: guests may now download files from protected directories
* added: LEPTON 2.x I18n language support is right now implemented in kitIdea
* changed: separated language depending settings from the language files
* added: support for english language
* added: calculate table cells with SUM, ADD, SUB and DIV (see documentation)

Release 0.21 - 2011-11-25

* fixed: missing command parameter in class.cronjob.php disturb links in reports
* fixed: duplicate entry for moving articles in group permissions
* added: select and restore articles by revisons - need group permission.

Release 0.20 - 2011-11-23

* added: fields "abstract", "minor changes" and "description" for articles
* changed: content and design of all reports (immediate, daily, weekly, monthly)
* added: project protocol show the last changes in a own tab - need permission
  and can be switched with the droplet parameter "section_protocol"

Release 0.19 - 2011-11-20

* added: parameter "log=login" in droplet "kit_idea" tracks the login of kitIdea
  users in the KIT account protocol 

Release 0.18 - 2011-11-16

* fixed: visitors can't see public projects
* added: can change email reporting in user settings now for all project groups
  at once

Release 0.17 - 2011-11-15

* changed: kitIdea uses now LibraryAdmin for including jQuery
* fixed: drag & drop problem at the project overview
* added: immediated, daily, weekly and monthly reports. User can switch between
  the report modes by links
* changed: mail.xxx.lte templates replaced by status.xxx.lte templates

Release 0.16 - 2011-11-14

* changed: kitIdea process status mails now with the KIT cronjob.

Release 0.15 - 2011-10-21

* added: kitIdea can calculate within HTML tables sums and multiplications

Release 0.14 - 2011-10-15

* changed: kitIdea is now using project-groups and allows multiple usage at 
  different pages within the same webspace
* added: comprehensive access-rights allow to define up to 5 access-groups at each
  project-group, specify rights for the project, sections, articles, files a.s.o.
* added: default access-rights for just registered users
* changed: now uses the CKE as editor and no longer the WB/LEPTON default editor
* changed: CKEditor is restricted to the defined project space within the MEDIA
  directory, users can no longer access to files of other projects and system
* added: features of the CKEditer are controlled via access-rights of the group
* changed: inbound kitDirList is now restricted to the defined space for each 
  project - user can no longer access to files of other projects
* changed: features and rights within kitDirList are now controlled via the group
  access-rights of the projects
* improved: inbound features of kitDirList within kitIdea
* added: drag & drop sorting of projects within a project group 

Release 0.13 - 2011-09-16

* added: parameter lepton_groups for droplet kit_idea allows to login via LEPTON 
  user interface. The user must be member of one of the specified groups.

Release 0.12 - 2011-09-13

* fixed: missing the section identifier at creating new articles moves the 
  articles to an non existing section

Release 0.11 - 2011-09-12

* fixed: moving articles between sections does not work
* fixed: missing check for dropletsExtension in precheck.php 
* fixed: some minor problems, code cleanup
* added: email information on project changes for all subscribed persons

Release 0.10 - 2011-09-11

* First Beta Release

upgrade.php
<?php

/**
 * kitIdea
 *
 * @author Ralf Hertsch <ralf.hertsch@phpmanufaktur.de>
 * @link http://phpmanufaktur.de
 * @copyright 2011 - 2012
 * @license MIT License (MIT) http://www.opensource.org/licenses/MIT
 */

// include class.secure.php to protect this file and the whole CMS!
if (defined('WB_PATH')) {
  if (defined('LEPTON_VERSION'))
    include(WB_PATH.'/framework/class.secure.php');
}
else {
  $oneback = "../";
  $root = $oneback;
  $level = 1;
  while (($level < 10) && (!file_exists($root.'/framework/class.secure.php'))) {
    $root .= $oneback;
    $level += 1;
  }
  if (file_exists($root.'/framework/class.secure.php')) {
    include($root.'/framework/class.secure.php');
  }
  else {
    trigger_error(sprintf("[ <b>%s</b> ] Can't include class.secure.php!", $_SERVER['SCRIPT_NAME']), E_USER_ERROR);
  }
}
// end include class.secure.php

// use LEPTON 2.x I18n for access to language files
if (! class_exists('LEPTON_Helper_I18n')) require_once WB_PATH . '/modules/' . basename(dirname(__FILE__)) . '/framework/LEPTON/Helper/I18n.php';
global $I18n;
if (!is_object($I18n)) {
    $I18n = new LEPTON_Helper_I18n();
}
else {
    $I18n->addFile('DE.php', WB_PATH.'/modules/'.basename(dirname(__FILE__)).'/languages/');
}

// load language depending onfiguration
if (!file_exists(WB_PATH.'/modules/' . basename(dirname(__FILE__)) . '/languages/' . LANGUAGE . '.cfg.php')) {
    require_once(WB_PATH .'/modules/'.basename(dirname(__FILE__)).'/languages/DE.cfg.php');
} else {
    require_once(WB_PATH .'/modules/'.basename(dirname(__FILE__)).'/languages/' .LANGUAGE .'.cfg.php');
}
if (! file_exists(WB_PATH . '/modules/' . basename(dirname(__FILE__)) . '/languages/' . LANGUAGE . '.php')) {
    if (! defined('KIT_FORM_LANGUAGE')) define('KIT_IDEA_LANGUAGE', 'DE'); // important: language flag is used by template selection
} else {
    if (! defined('KIT_FORM_LANGUAGE')) define('KIT_IDEA_LANGUAGE', LANGUAGE);
}

// ACHTUNG: EINBINDUNG NUR NOCH VORLAUEFIG BIS I18n VOLLSTAENDIG !!!
if(!file_exists(WB_PATH .'/modules/'.basename(dirname(__FILE__)).'/languages/' .LANGUAGE .'.php')) {
	require_once(WB_PATH .'/modules/'.basename(dirname(__FILE__)).'/languages/DE.php'); // Vorgabe: DE verwenden
}
else {
	require_once(WB_PATH .'/modules/'.basename(dirname(__FILE__)).'/languages/' .LANGUAGE .'.php');
}

require_once WB_PATH.'/modules/kit_tools/class.droplets.php';
require_once WB_PATH.'/modules/'.basename(dirname(__FILE__)).'/class.idea.php';
require_once WB_PATH.'/modules/kit_form/class.form.php';


global $admin;
global $database;

$error = '';

function getRelease() {
    // read info.php into array
    $info_text = file(WB_PATH.'/modules/'.basename(dirname(__FILE__)).'/info.php');
    if ($info_text == false) {
        return -1;
    }
    // walk through array
    foreach ($info_text as $item) {
        if (strpos($item, '$module_version') !== false) {
            // split string $module_version
            $value = explode('=', $item);
            // return floatval
            return floatval(preg_replace('([\'";,\(\)[:space:][:alpha:]])', '', $value[1]));
        }
    }
    return -1;
} // getVersion()
$release = sprintf('%01.2f', getRelease());

// install missing tables
$tables = array(
        'dbIdeaCfg',
        'dbIdeaProject',
        'dbIdeaProjectSections',
        'dbIdeaProjectArticles',
        'dbIdeaRevisionArchive',
        'dbIdeaTableSort',
        'dbIdeaProjectGroups',
        'dbIdeaProjectAccess',
        'dbIdeaProjectUsers',
        'dbIdeaStatusChange');

foreach ($tables as $table) {
	$create = null;
	$create = new $table();
	if (!$create->sqlTableExists()) {
		if (!$create->sqlCreateTable()) {
			$error .= sprintf('[INSTALLATION %s] %s', $table, $create->getError());
		}
	}
}

// special for Release 0.18
if ($release == '0.18') {
    $dbIdeaProjectUsers = new dbIdeaProjectUsers();
    $where = array(dbIdeaProjectUsers::field_status => dbIdeaProjectUsers::status_active);
    $data = array(dbIdeaProjectUsers::field_email_info => dbIdeaProjectUsers::EMAIL_DAILY);
    if (!$dbIdeaProjectUsers->sqlUpdateRecord($data, $where)) {
        $error .= sprintf('[UPGRADE] Operation fail: %s', $dbIdeaProjectUsers->getError());
    }
}

// Release 0.20
$dbIdeaProjectArticles = new dbIdeaProjectArticles();
if (!$dbIdeaProjectArticles->sqlFieldExists(dbIdeaProjectArticles::field_abstract)) {
    // added abstract for changes and change type
    if (!$dbIdeaProjectArticles->sqlAlterTableAddField(dbIdeaProjectArticles::field_abstract, "VARCHAR(255) NOT NULL DEFAULT ''", dbIdeaProjectArticles::field_kit_contact_id)) {
        $error .= sprintf('[UPGRADE mod_kit_idea_project_articles] %s', $dbIdeaProjectArticles->getError());
    }
    if (!$dbIdeaProjectArticles->sqlAlterTableAddField(dbIdeaProjectArticles::field_description, "VARCHAR(512) NOT NULL DEFAULT ''", dbIdeaProjectArticles::field_kit_contact_id)) {
        $error .= sprintf('[UPGRADE mod_kit_idea_project_articles] %s', $dbIdeaProjectArticles->getError());
    }
    if (!$dbIdeaProjectArticles->sqlAlterTableAddField(dbIdeaProjectArticles::field_change, "TINYINT NOT NULL DEFAULT '".dbIdeaProjectArticles::CHANGE_NORMAL."'", dbIdeaProjectArticles::field_kit_contact_id)) {
        $error .= sprintf('[UPGRADE mod_kit_idea_project_articles] %s', $dbIdeaProjectArticles->getError());
    }
}

$dbIdeaProject = new dbIdeaProject();
if (!$dbIdeaProject->sqlFieldExists(dbIdeaProject::field_url)) {
    // add URL field
    if (!$dbIdeaProject->sqlAlterTableAddField(dbIdeaProject::field_url, "VARCHAR(255) NOT NULL DEFAULT ''", dbIdeaProject::field_status)) {
        $error .= sprintf('[UPGRADE mod_kit_idea_project] %s', $dbIdeaProject->getError());
    }
}

// Release 0.22
if (!$dbIdeaProject->sqlAlterTableChangeField(dbIdeaProject::field_desc_short, dbIdeaProject::field_desc_short, "TEXT NOT NULL DEFAULT ''")) {
    $error .= sprintf('[UPGRADE mod_kit_idea_project] %s', $dbIdeaProject->getError());
}



// import forms from /forms to kitForm
$kitForm = new dbKITform();
$dir_name = WB_PATH.'/modules/'.basename(dirname(__FILE__)).'/forms/';
$folder = opendir($dir_name);
$names = array();
while (false !== ($file = readdir($folder))) {
    $ff = array();
    $ff = explode('.', $file);
    $ext = end($ff);
    if ($ext	==	'kit_form') {
        $names[] = $file;
    }
}
closedir($folder);
$message = '';
foreach ($names as $file_name) {
    $form_file = $dir_name.$file_name;
    $form_id = -1;
    $msg = '';
    if (!$kitForm->importFormFile($form_file, '', $form_id, $msg, true)) {
        if ($kitForm->isError()) $error .= sprintf('[IMPORT FORM %s] %s', $file_name, $kitForm->getError());
    }
    $message .= $msg;
}

// remove Droplets
$dbDroplets = new dbDroplets();
// the array contains the droplets to remove
$droplets = array('kit_idea');
foreach ($droplets as $droplet) {
    $where = array(dbDroplets::field_name => $droplet);
    if (!$dbDroplets->sqlDeleteRecord($where)) {
        $message .= sprintf('[UPGRADE] Error uninstalling Droplet: %s', $dbDroplets->getError());
    }
}

// Install Droplets
$droplets = new checkDroplets();
$droplets->droplet_path = WB_PATH.'/modules/kit_idea/droplets/';

if ($droplets->insertDropletsIntoTable()) {
    $message .= 'The droplets for kitIdea are successfully installed.\n\nPlease check the kitIdea history for changes in the program and for changes in the templates (this is very important if you are using user defined templates!).';
}
else {
    $message .= sprintf('Error installing the droplets for kitIdea: %s', $droplets->getError());
}
if ($message != "") {
    echo '<script language="javascript">alert ("'.$message.'");</script>';
}

// delete no longer needed files
$delete_files = array(
        'kit_idea.js',
        'kit_idea.preset',
        'class.status.mail.php',
        'templates/1/DE/mail.article.inserted.lte',
        'templates/1/DE/mail.article.update.lte',
        'templates/1/DE/mail.project.inserted.lte',
        'templates/1/DE/mail.project.updated.lte'
);
foreach ($delete_files as $file) {
    if (file_exists(WB_PATH.'/modules/kit_idea/'.$file)) {
        @unlink(WB_PATH.'/modules/kit_idea/'.$file);
    }
}


// Prompt Errors
if (!empty($error)) {
	$admin->print_error($error);
}